### PR TITLE
Parallel Cross Entropy using online softmax

### PR DESCRIPTION
**Pull request by sanandaraj5597**
# Description

This PR implements a parallel cross entropy function using the online technique to calculate softmax. This feature has multiple aspects:

1) The vocab dimension can be sharded along the TP axis to perform this loss calculation in a distributed fashion.
2) Online softmax helps us parallelize the softmax calculation giving us more efficiency.
3) Calculating gradients in the forward itself, so backward step is a no-op. 
4) Storing the gradients in-place of the input tensor saving memory.
5) OAI Triton implementation helps us integrate GPU kernel level semantics and torch level communication API's together.

[Thanks to Liger kernel implementation for providing the idea about online softmax and in-place gradient calculation.]

### CODE REVIEW

**Review by timmoon10 — COMMENTED** [review]
Can you add a test in [`tests/pytorch`](https://github.com/NVIDIA/TransformerEngine/tree/main/tests/pytorch)?

**Comment by sanandaraj5597** [comment]
@timmoon10 Added tests.

File: transformer_engine/pytorch/triton/cross_entropy.py [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+
+from typing import Union
+from functools import reduce
+from operator import mul
+
+import torch
+import torch.distributed as dist
+
+import triton
+import triton.language as tl
```
Comment (ksivaman) [inline transformer_engine/pytorch/triton/cross_entropy.py:15]: Even though triton is a dependency for most torch builds, it's a good idea to explicitly specify this as a dependency in our `setup.py`.

File: tests/pytorch/test_parallel_cross_entropy.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
+import random
```
Comment (ksivaman) [inline tests/pytorch/test_parallel_cross_entropy.py:1]: Currently this tests is unused, we should ensure to call this in `qa/L0_pytorch_unittest/test.sh`.

File: transformer_engine/pytorch/triton/cross_entropy.py (hunk 1 above)
Comment (sanandaraj5597) [reply transformer_engine/pytorch/triton/cross_entropy.py:15]: Can you tell me how to add this?

File: tests/pytorch/test_parallel_cross_entropy.py (hunk 2 above)
Comment (sanandaraj5597) [reply tests/pytorch/test_parallel_cross_entropy.py:1]: Added

File: transformer_engine/pytorch/triton/cross_entropy.py (hunk 1 above)
Comment (sanandaraj5597) [reply transformer_engine/pytorch/triton/cross_entropy.py:15]: Done. Thanks!

**Review by ksivaman — COMMENTED** [review]
@sanandaraj5597 Could you fix the linting errors? You could run it locally using `bash qa/L0_pytorch_lint/test.sh`

**Comment by sanandaraj5597** [comment]
Fixed lint errors.

File: setup.py [hunk 3]
```diff
@@ -104,6 +104,7 @@ def setup_requirements() -> Tuple[List[str], List[str], List[str]]:
     if not bool(int(os.getenv("NVTE_RELEASE_BUILD", "0"))):
         if "pytorch" in frameworks:
             install_reqs.extend(["torch"])
+            install_reqs.extend(["triton"])
```
Comment (ksivaman) [inline setup.py:107]:
```suggestion
            install_reqs.extend(["torch", "triton"])
```

**Review by timmoon10 — APPROVED** [review]
LGTM, pending CI

**Review by ksivaman — APPROVED** [review]
LGTM!

**Comment by timmoon10** [comment]
It seems that the Triton dependency is messing up other tests. The latest upstream Triton (3.2.0) doesn't support Blackwell, so the NVIDIA PyTorch container is using a custom internal build. I've removed Triton as a formal dependency, but we should put it back once Blackwell support is upstreamed.
